src/Metrics.Extensions/src/Internal/Microsoft.ApplicationInsights.Metrics/GaugeAggregator.cs

```diff
@@ -0,0 +1,156 @@
+                                                DataSeries?.MetricIdentifier.MetricId ?? Util.NullString,
+                                                MetricConfigurations.Common.Gauge().Constants().AggregateKindMoniker);
+
+            aggregate.Data[MetricConfigurations.Common.Gauge().Constants().AggregateKindDataKeys.Last] = last;
+            aggregate.Data[MetricConfigurations.Common.Gauge().Constants().AggregateKindDataKeys.Min] = min;
+            aggregate.Data[MetricConfigurations.Common.Gauge().Constants().AggregateKindDataKeys.Max] = max;
+
+            AddInfo_Timing_Dimensions_Context(aggregate, periodEnd);
+
+            return aggregate;
+        }
+
+        protected override void ResetAggregate()
+        {
+            lock (_updateLock)
+            {
+                _data.Min = Double.MaxValue;
+                _data.Max = Double.MinValue;
+                _data.Last = Double.NaN;
+                _data.HasValues = false;
+            }
+        }
+
+        protected override object UpdateAggregate_Stage1(MetricValuesBufferBase<double> buffer, int minFlushIndex, int maxFlushIndex)
+        {
+            // Compute a summary of the buffer:
+            Data bufferData = new Data();
+            bufferData.HasValues = false;
+            bufferData.Min = Double.MaxValue;
+            bufferData.Max = Double.MinValue;
+            bufferData.Last = Double.NaN;
+
+            for (int index = minFlushIndex; index <= maxFlushIndex; index++)
+            {
+                double metricValue = buffer.GetAndResetValue(index);
+
+                if (Double.IsNaN(metricValue))
+                {
+                    continue;
+                }
+
+                bufferData.HasValues = true;
+                bufferData.Max = (metricValue > bufferData.Max) ? metricValue : bufferData.Max;
+                bufferData.Min = (metricValue < bufferData.Min) ? metricValue : bufferData.Min;
+                bufferData.Last = metricValue;
+            }
+
+            return bufferData;
+        }
+
+        protected override void UpdateAggregate_Stage2(object stage1Result)
+        {
+            Data bufferData = (Data) stage1Result;
+
+            if (! bufferData.HasValues)
+            {
+                return;
+            }
+
+            if (_restrictToUInt32Values)
+            {
+                bufferData.Max = Math.Round(bufferData.Max);
+                bufferData.Min = Math.Round(bufferData.Min);
+                bufferData.Last = Math.Round(bufferData.Last);
+            }
+
+            // Take a lock and update the aggregate:
+            lock (_updateLock)
+            {
+                _data.HasValues = true;
+                _data.Max = (bufferData.Max > _data.Max) ? bufferData.Max : _data.Max;
+                _data.Min = (bufferData.Min < _data.Min) ? bufferData.Min : _data.Min;
+                _data.Last = bufferData.Last;
+            }
+        }
+
+#pragma warning disable SA1401 // Fields must be private
```

Comment:
why wouldn't you create properties instead of disabling the warning? Or make it a `struct`?

Reply:
Just an encapsulation private to this aggregator. Minimizing code. struct would make it less efficient for passing around.

Reply:
So why not properties? Compiler will optimize everything out